### PR TITLE
Use a better data structure for VM compilation of free vars.

### PR DESCRIPTION
**Pull request by ppedrot**
This fixes #3450 and probably provides a huge speed-up to many instances.

### CODE REVIEW

**Comment by ppedrot** [comment]
People interested may include @bgregoir, @mattam82 and @JasonGross.

**Comment by silene** [comment]
Do not you want to remove the `find_at` function? Because the code is looking really dumb now:

    let find_at fv env =
      let n = FvMap.find fv env.fv_fwd in
      env.size - n
    ...
    env.size - (find_at cid env)

**Comment by mattam82** [comment]
LGTM. I'm curious of this implementation that looks for variable decls in the env and might not find them there, it's trying to build the smallest closure environments possible?

**Comment by ppedrot** [comment]
@silene: here is a slightly cleaned-up version as you suggested.

**Comment by silene** [comment]
As far as I am concerned, I don't know what could go wrong (except for the comparison function, which I double-checked).

**Comment by ppedrot** [comment]
I think that a positive opinion from two core-devs for a near-trivial change is enough. Can somebody merge it? It may be good practice to let others perform the merge in kernel patches, it sounds like a moral engagement :)

BTW, what about backporting it into 8.5? It is quite critical for universe polymorphic code using vm_compute AFAIU.

File: kernel/cbytecodes.ml [hunk 1]
```diff
@@ -142,11 +142,32 @@ type fv = fv_elem array
 exception NotClosed
 
 
+module Fv_elem =
+struct
+type t = fv_elem =
+| FVnamed of Id.t
+| FVrel of int
+| FVuniv_var of int
```
Comment (aspiwack) [inline kernel/cbytecodes.ml]: Repeating the constructors, here, does not look useful. I would remove them.

**Comment by ppedrot** [comment]
Added the change suggested by @aspiwack .

**Comment by maximedenes** [comment]
This change, in particular the small arithmetic on indices, seems correct.

**Pull request merged**